BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,175 @@
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature == BL_PLD_COMM_SIG) {
+      Ptr += ((BL_PLD_COMM_HDR *)Ptr)->TotalSize;
+      continue;
+    }
+    break;
+  }
+
+  if ( Ptr + ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize > SmmBase + SmmSize ) {
+    DEBUG ((DEBUG_ERROR, "Overflowing SMM region\n"));
+    return;
+  }
+  CopyMem ((VOID *)Ptr, DataPtr, ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize);
+}
+
+
+/**
+  Get pointer to a specific struct area in TSEG
+
+  @param      Id                Id of the struct to be retreived
+
+  @retval     NULL              Struct with signature not found
+  @retval     Ptr               Pointer to struct with signature
+**/
+VOID *
+EFIAPI
+GetS3Info (
+  IN  UINT8   Id
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+  BOOLEAN           FoundInfo = FALSE;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return NULL;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature == BL_PLD_COMM_SIG) {
+      if ( ((BL_PLD_COMM_HDR *)Ptr)->Id != Id) {
+        Ptr += ((BL_PLD_COMM_HDR *)Ptr)->TotalSize;
+        continue;
+      }
+      FoundInfo = TRUE;
+      break;
+    }
+  }
+
+  if (!FoundInfo) {
+    return NULL;
+  }
+
+  return (VOID *) Ptr;
+}
+
+
+/**
+  This function restores the states of the registers that were
+  set to be saved by the bootloader in the normal boot path.
+  After the payload has updated the 'Val' field of the reg info
+  in the normal boot path, we simply OR the updated 'Val'
+  with the existing vale of the register in the S3 resume boot path.
+  This function is only called in the S3 resume path.
+
+  @param    S3SaveReg   S3_SAVE_REG info offset
+    
+  @retval   None
+
+**/
+VOID
+EFIAPI
+RestoreS3RegInfo (
+  IN  S3_SAVE_REG   *S3SaveReg
+  )
+{
+  UINT8                 Index;
+  UINT32                Data32;
+  UINT16                Data16;
+
+  if (S3SaveReg == NULL || S3SaveReg->S3SaveHdr.Id != S3_SAVE_REG_COMM_ID) {
+    DEBUG ((DEBUG_ERROR, "Invalid S3 Save Reg pointer\n"));
+    return;
+  }
+
+  for (Index = 0; Index < S3SaveReg->S3SaveHdr.Count; Index++) {
+    if (S3SaveReg->RegInfo[Index].Addr != 0x00) {
+      //
+      // Check register type and width
+      //
+      if ( ((S3SaveReg->RegInfo[Index].Type != IO) && (S3SaveReg->RegInfo[Index].Type != MEM)) ||
+           ((S3SaveReg->RegInfo[Index].Width != WIDE32) && (S3SaveReg->RegInfo[Index].Width != WIDE16)) ) {
```

Comment:
Any reason why  WIDE8 is not supported ?

Reply:
No specific reason, we can add support of 8-bit wide registers if wee need to save one such reg. Currently, we are only saving 32-bit wide registers.